m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,781 @@
+import com.uber.m3.tally.m3.thrift.TMultiUdpClient;
+import com.uber.m3.tally.m3.thrift.TUdpClient;
+import com.uber.m3.thrift.generated.CountValue;
+import com.uber.m3.thrift.generated.GaugeValue;
+import com.uber.m3.thrift.generated.M3;
+import com.uber.m3.thrift.generated.Metric;
+import com.uber.m3.thrift.generated.MetricBatch;
+import com.uber.m3.thrift.generated.MetricTag;
+import com.uber.m3.thrift.generated.MetricValue;
+import com.uber.m3.thrift.generated.TimerValue;
+import com.uber.m3.util.Duration;
+import com.uber.m3.util.ImmutableMap;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.protocol.TProtocolFactory;
+import org.apache.thrift.transport.TTransport;
+
+import java.net.InetAddress;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Phaser;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.thrift.transport.TTransportException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An M3 implementation of a {@link StatsReporter}.
+ */
+public class M3Reporter implements StatsReporter, AutoCloseable {
+    public static final String SERVICE_TAG = "service";
+    public static final String ENV_TAG = "env";
+    public static final String HOST_TAG = "host";
+    public static final String DEFAULT_TAG_VALUE = "default";
+
+    public static final String DEFAULT_HISTOGRAM_BUCKET_ID_NAME = "bucketid";
+    public static final String DEFAULT_HISTOGRAM_BUCKET_NAME = "bucket";
+    public static final int DEFAULT_HISTOGRAM_BUCKET_TAG_PRECISION = 6;
+
+    private static final Logger LOG = LoggerFactory.getLogger(M3Reporter.class);
+
+    private static final int MAX_PROCESSOR_WAIT_ON_CLOSE_MILLIS = 1000;
+    private static final int MAX_PROCESSOR_WAIT_UNTIL_FLUSH_MILLIS = 1000;
+
+    private static final int DEFAULT_FLUSH_TRY_COUNT = 2;
+    private static final int FLUSH_RETRY_DELAY_MILLIS = 200;
+
+    private static final int DEFAULT_METRIC_SIZE = 100;
+
+    private static final int DEFAULT_MAX_QUEUE_SIZE = 4096;
+    private static final int DEFAULT_MAX_PACKET_SIZE = 1440;
+
+    private static final int THREAD_POOL_SIZE = 10;
+
+    private static final int EMIT_METRIC_BATCH_OVERHEAD = 19;
+    private static final int MIN_METRIC_BUCKET_ID_TAG_LENGTH = 4;
+
+    private M3.Client client;
+
+    private final MetricBatch metricBatch = new MetricBatch();
+
+    private final Object calcLock = new Object();
```

Comment:
just occurred to me, can't you use the object itself for locking? as in why have an additional member called `calcLock` when you can synchronise on `calc`

Reply:
`calcLock` protects both `calc` and `calcProtocol`. I'll put in a comment.